File: javascript_code/highlevel/heap/DataStream.js
// 抖码算法，让算法学习变的简单有趣
// 作者：老汤


class Node {
    constructor(val) {
        this.val = val
        this.next = null
    }
}

class DataStream {
    constructor() {
        this.dummyNode = new Node(-1)
    }

    // O(1)
    add(val) {
        const newNode = new Node(val)
        newNode.next = this.dummyNode.next
        this.dummyNode.next = newNode
    }

    // O(n)
    removeMax() {
        if (!this.dummyNode.next) {
            throw new Error("removeMax 失败，因为数据流中没有元素")
        }
        // 遍历一遍链表，找到最大值所在的节点
        let curr = this.dummyNode.next
        let maxValueNode = curr
        while (curr) {
            if (curr.val > maxValueNode.val) {
                maxValueNode = curr
            }
            curr = curr.next
        }

        // 找到最大值节点的前一个节点
        curr = this.dummyNode.next
        let maxValueNodePrev = this.dummyNode
        while (curr) {
            if (curr == maxValueNode) {
                break;
            }
            maxValueNodePrev = curr
            curr = curr.next
        }

        // 删除最大值所在的节点
        maxValueNodePrev.next = maxValueNode.next
        maxValueNode.next = null

        return maxValueNode.val
    }
}

function test() {
    const dataStream = new DataStream()
    dataStream.add(3)
    console.log(dataStream.removeMax()) // 打印 3
    dataStream.add(6)
    dataStream.add(5)
    console.log(dataStream.removeMax()) // 打印 6
    dataStream.add(2)
    dataStream.add(1)
    console.log(dataStream.removeMax()) // 打印 5
}

test()

File: python_code/highlevel/heap/DataStream.py
# 抖码算法，让算法学习变得简单有趣
# 作者：老汤


class Node:
    def __init__(self, val, next=None):
        self.val = val
        self.next = next


class DataStream:
    def __init__(self):
        self.dummy_node = Node(-1)

    # 时间复杂度：O(1)
    def add(self, val):
        new_node = Node(val)
        new_node.next = self.dummy_node.next
        self.dummy_node.next = new_node

    def removeMax(self):
        if not self.dummy_node.next:
            raise Exception("removeMax 失败，因为数据流中没有元素")
        # 遍历一遍链表，找到最大值所在的节点
        curr = self.dummy_node.next
        max_value_node = curr
        while curr:
            if curr.val > max_value_node.val:
                max_value_node = curr
            curr = curr.next

        # 找到最大值节点的前一个节点
        curr, max_value_node_prev = self.dummy_node.next, self.dummy_node
        while curr:
            if curr == max_value_node:
                break
            max_value_node_prev = curr
            curr = curr.next

        # 删除最大值所在的节点
        max_value_node_prev.next = max_value_node.next
        max_value_node.next = None

        return max_value_node.val


if __name__ == '__main__':
    ds = DataStream()
    ds.add(3)
    print(ds.removeMax())  # 打印 3
    ds.add(6)
    ds.add(5)
    print(ds.removeMax())  # 打印 6
    ds.add(2)
    ds.add(1)
    print(ds.removeMax())  # 打印 5
